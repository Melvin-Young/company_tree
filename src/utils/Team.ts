export interface IStaff {
  id: number,
  name: string,
  team: string
}

export interface ITeam {
  name: string,
  getParent: () => ITeam,
  setParent: (parent: ITeam) => void,
  getChildren: () => Map<string, ITeam>,
  getChild: (child: string) => ITeam,
  addChild: (child: ITeam) => Map<string, ITeam>,
  hasChild: (child: string) => boolean,
  getStaff: () => Map<number, IStaff>,
  addStaff: (staff: IStaff) => Map<number, IStaff>,
  getNestedStaff: () => Map<number, IStaff>,
  addNestedStaff: (staff: IStaff) => Map<number, IStaff>
}

export default class Team implements ITeam {
  public name: string;
  private parent: ITeam;
  private children: Map<string, ITeam>;
  private staff: Map<number, IStaff>;
  private nestedStaff: Map<number, IStaff>;

  constructor(name: string) {
    if (!name) {
      throw new Error('Name is required');
    }
    this.name = name;
    this.parent = {} as ITeam;
    this.children = new Map();
    this.staff = new Map();
    this.nestedStaff = new Map();
  }

  public getParent(): ITeam {
    return this.parent;
  }

  public setParent(parent: ITeam): void {
    this.parent = parent;
  }

  public getChildren(): Map<string, ITeam> {
    return this.children;
  }

  public getChild(child: string): ITeam {
    return this.children.get(child) as ITeam;
  }

  public addChild(child: ITeam): Map<string, ITeam> {
    return this.children.set(child.name, child);
  }

  public hasChild(child: string): boolean {
    return this.children.has(child);
  }

  public getStaff(): Map<number, IStaff> {
    return this.staff;
  }

  public addStaff(staff: IStaff): Map<number, IStaff> {
    return this.staff.set(staff.id, staff);
  }

  public getNestedStaff(): Map<number, IStaff> {
    return this.nestedStaff;
  }

  public addNestedStaff(staff: IStaff): Map<number, IStaff> {
    return this.nestedStaff.set(staff.id, staff);
  }
}