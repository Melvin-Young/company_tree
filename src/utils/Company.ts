import Team, {IStaff, ITeam} from './Team';

export interface ICompany {
  companyName: string,
  getRoot: () => ITeam,
  addTeam: (team: ITeam, ancestors: string[]) => Map<string, ITeam>,
  search: (ancestors: string[]) => ITeam,
  addMemberToTeam: (pathToTeam: string[], staff: IStaff) => Map<number, IStaff>,
  getStaff: (pathToTeam: string[]) => Map<number, IStaff>,
}

export default class Company implements ICompany {
  public companyName: string;

  private root: ITeam;

  constructor(name: string) {
    this.companyName = name;
    this.root = new Team(name);
  }

  public getRoot(): ITeam {
    // Returns iterator for the values of the base node. 
    // It only ever has one node, which is the top level of the company
    return this.root;

  }

  public addTeam(team: ITeam, ancestors: string[]): Map<string, ITeam> {
    // Searches through list of ancestor teams back to root to make sure team is being added to connected tree
    // If is the first child then appends directly to root company node
    if(!this.root.getChildren().size) {
      return this.root.addChild(team);
    }

    const parent = this.search(ancestors);
    team.setParent(parent);
    return parent.addChild(team);
  }

  public search(pathToTeam: string[]): ITeam {
    // Searches for nodes by checking each nodes Map() of its children for the next value in the list
    // If found it returns the object and then checks that object children for the next node in the path
    let currentAncestor: ITeam = this.root;
    let nextAncestor = pathToTeam.shift();

    while(nextAncestor && currentAncestor.hasChild(nextAncestor)) {
      currentAncestor = currentAncestor.getChild(nextAncestor);
      if (!pathToTeam.length) {
        return currentAncestor;
      }
      nextAncestor = pathToTeam.shift();
    }
    throw new Error('Not found');
  }

  public addMemberToTeam(pathToTeam: string[], staff: IStaff): Map<number, IStaff> {
    // Rehashed search function adds staff member to each node enroute to finding the node representing the team the staff member belongs to
    // Can help avoid expense of having to visit each nested node later
    let currentAncestor: ITeam = this.root;
    let nextAncestor = pathToTeam.shift();

    while(nextAncestor && currentAncestor.hasChild(nextAncestor)) {
      currentAncestor = currentAncestor.getChild(nextAncestor);
      // Had trouble merging maps so adding immediate and nested staff to nested staff Map() until time for a fix
      currentAncestor.addNestedStaff(staff);

      if (!pathToTeam.length) {
        currentAncestor.addStaff(staff);
        return currentAncestor.getStaff();
      }

      nextAncestor = pathToTeam.shift();
    }
    throw new Error('Not found');
  }

  public getStaff(pathToTeam: string[]): Map<number, IStaff> {
    return this.search(pathToTeam).getNestedStaff();
    
  }
}
