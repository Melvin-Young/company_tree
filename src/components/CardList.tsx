/* tslint:disable: no-unused-expression jsx-no-lambda*/

import React, { Component} from 'react';
import { ICompany } from '../utils/Company';
import { ITeam } from '../utils/Team';
import { IStaff } from '../utils/Team';
import HeaderCard from './HeaderCard';
import StaffCard from './StaffCard';
import TeamCard from './TeamCard';

interface IProps {
  company: ICompany,
  renderTarget: ITeam,
  setHeader: (team: ITeam) => void
}

interface IState {
  renderTarget: ITeam,
  showStaff: boolean
}
class CardList extends Component<IProps, IState> {
  constructor(props: IProps) {
    super(props);
    this.state = {
      renderTarget: props.renderTarget,
      showStaff: false
    }
  }

  // Updates state whenever a card is clicked and new props for the team to render are passed down
  public componentDidUpdate(prevProps: IProps): void {
    this.props !== prevProps && this.setState({renderTarget: this.props.renderTarget});
  }

  // Show Staff and Show Teams toggle whether Team Cards or Staff Cards will be rendered in body of component
  public showStaff = () => {
    this.setState({showStaff: true});
  }

  public showTeams = () => {
    this.setState({showStaff: false});
  }

  public render() {
    const { setHeader, company } = this.props;
    const teamList: JSX.Element[] = [];
    const staffList: JSX.Element[] = [];
    
    this.state.renderTarget.getChildren().forEach((childTeam: ITeam) => {
      return teamList.push( 
        <TeamCard
          company={company}
          key={childTeam.name}
          setHeader={setHeader}
          team={childTeam}/>
        );
    });

    // Nested Staff is all staff assigned to a team or assigned to one of its sub teams
    this.state.renderTarget.getNestedStaff().forEach((staffMember: IStaff) => {
      return staffList.push(<StaffCard key={staffMember.id} staff={staffMember}/>);
    });

    return (
      <section>
        <div className="card-container-banner">
          <HeaderCard
            key={this.state.renderTarget.name}
            setHeader={setHeader}
            triggerShowTeams = {this.showTeams}
            triggerShowStaff = {this.showStaff}
            team={this.state.renderTarget}/>
        </div>
        <hr/>
        <div className="card-container">
          { this.state.showStaff ? staffList : teamList }
        </div>
      </section>
    );
  }
}
export default CardList;