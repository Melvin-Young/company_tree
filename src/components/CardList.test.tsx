/* tslint:disable no-unused-expression jsx-no-lambda */

import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';

import Company from '../utils/Company';
import Team from '../utils/Team';
import CardList from './CardList';

Enzyme.configure({ adapter: new Adapter()});

const dummyCompany = new Company('Dummy');
const dummyTeam = new Team('Dummy Team');
dummyCompany.addTeam(dummyTeam, ['Dummy']);

describe('TeamCard', () => {
    it('should render', () => {
        const component = shallow(<CardList renderTarget={dummyTeam} company={dummyCompany} setHeader={jest.fn()} />)
        expect(component).toMatchSnapshot();
    });
});
