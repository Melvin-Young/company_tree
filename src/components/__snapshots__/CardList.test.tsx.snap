// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamCard should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CardList
    company={
        Company {
            "companyName": "Dummy",
            "root": Team {
              "children": Map {
                "Dummy Team" => Team {
                  "children": Map {},
                  "name": "Dummy Team",
                  "nestedStaff": Map {},
                  "parent": Object {},
                  "staff": Map {},
                },
              },
              "name": "Dummy",
              "nestedStaff": Map {},
              "parent": Object {},
              "staff": Map {},
            },
          }
    }
    renderTarget={
        Team {
            "children": Map {},
            "name": "Dummy Team",
            "nestedStaff": Map {},
            "parent": Object {},
            "staff": Map {},
          }
    }
    setHeader={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="card-container-banner"
>
          <HeaderCard
                    setHeader={[Function]}
                    team={
                              Team {
                                        "children": Map {},
                                        "name": "Dummy Team",
                                        "nestedStaff": Map {},
                                        "parent": Object {},
                                        "staff": Map {},
                                      }
                    }
                    triggerShowStaff={[Function]}
                    triggerShowTeams={[Function]}
          />
</div>,
        <hr />,
        <div
          className="card-container"
>
          
</div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <HeaderCard
            setHeader={[Function]}
            team={
                        Team {
                                    "children": Map {},
                                    "name": "Dummy Team",
                                    "nestedStaff": Map {},
                                    "parent": Object {},
                                    "staff": Map {},
                                  }
            }
            triggerShowStaff={[Function]}
            triggerShowTeams={[Function]}
/>,
          "className": "card-container-banner",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": "Dummy Team",
          "nodeType": "class",
          "props": Object {
            "setHeader": [Function],
            "team": Team {
              "children": Map {},
              "name": "Dummy Team",
              "nestedStaff": Map {},
              "parent": Object {},
              "staff": Map {},
            },
            "triggerShowStaff": [Function],
            "triggerShowTeams": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "hr",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "card-container",
        },
        "ref": null,
        "rendered": Array [],
        "type": "div",
      },
    ],
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="card-container-banner"
>
            <HeaderCard
                        setHeader={[Function]}
                        team={
                                    Team {
                                                "children": Map {},
                                                "name": "Dummy Team",
                                                "nestedStaff": Map {},
                                                "parent": Object {},
                                                "staff": Map {},
                                              }
                        }
                        triggerShowStaff={[Function]}
                        triggerShowTeams={[Function]}
            />
</div>,
          <hr />,
          <div
            className="card-container"
>
            
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <HeaderCard
              setHeader={[Function]}
              team={
                            Team {
                                          "children": Map {},
                                          "name": "Dummy Team",
                                          "nestedStaff": Map {},
                                          "parent": Object {},
                                          "staff": Map {},
                                        }
              }
              triggerShowStaff={[Function]}
              triggerShowTeams={[Function]}
/>,
            "className": "card-container-banner",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": "Dummy Team",
            "nodeType": "class",
            "props": Object {
              "setHeader": [Function],
              "team": Team {
                "children": Map {},
                "name": "Dummy Team",
                "nestedStaff": Map {},
                "parent": Object {},
                "staff": Map {},
              },
              "triggerShowStaff": [Function],
              "triggerShowTeams": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "hr",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "card-container",
          },
          "ref": null,
          "rendered": Array [],
          "type": "div",
        },
      ],
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
